File: Darts/src/burlton/dartzee/code/screen/ai/VisualisationPanelDensity.java
package burlton.dartzee.code.screen.ai;

import java.awt.Color;
import java.awt.Point;
import java.awt.image.BufferedImage;
import java.util.HashMap;

import javax.swing.ImageIcon;
import javax.swing.JLabel;
import javax.swing.SwingConstants;

import burlton.dartzee.code.ai.AbstractDartsModel;
import burlton.dartzee.code.object.ColourWrapper;
import burlton.dartzee.code.utils.DartsColour;
import burlton.dartzee.code.utils.GeometryUtil;

public class VisualisationPanelDensity extends AbstractVisualisationPanel
{
	private static final int LABEL_WIDTH = 60;
	private static final int LABEL_HEIGHT = 30;
	
	private BufferedImage keyImg = null;
	
	public VisualisationPanelDensity()
	{
		super();
		
		panelKey.setBounds(0, 0, 100, 500);
		panel.add(panelKey);
		
		keyImg = new BufferedImage(panelKey.getWidth(), panelKey.getHeight(), BufferedImage.TYPE_INT_ARGB);
		panelKey.setIcon(new ImageIcon(keyImg));
	}
	
	private final JLabel panelKey = new JLabel();
	
	
	@Override
	public void showVisualisation(HashMap<Point, Integer> hmPointToCount, AbstractDartsModel model)
	{
		Point centerPt = model.getScoringPoint(dartboard);
		
		int width = overlay.getWidth();
		int height = overlay.getHeight();
		
		int[] pixels = new int[width * height];
		
		int i = 0;
		for (int y=0; y<height; y++)
		{
			for (int x=0; x<width; x++)
			{
		    	double radius = GeometryUtil.getDistance(new Point(x, y), centerPt);
		    	double probability = model.getProbabilityWithinRadius(radius);
		    	
		    	Color c = getColorForProbability(probability);
		    	pixels[i] = c.getRGB();
		    	i++;
			}
		}
		
		
		overlayImg.setRGB(0, 0, width, height, pixels, 0, width);
		repaint();
	}
	
	@Override
	public void paintKey()
	{
		int width = panelKey.getWidth();
		int height = panelKey.getHeight();
		
		int[] pixels = new int[width * height];
		
		//Label is 30 pixels wide. (100/2) - (50/2) = 25
		int lblXPosition = (panel.getWidth() / 2) - (LABEL_WIDTH/2);
		int i = 0;
		for (int y=0; y<height; y++)
		{
			for (int x=0; x<width; x++)
			{
		    	double probability = (float)y / (float)height;
		    	Color c = getColorForProbability(probability);
		    	
		    	pixels[i] = c.getRGB();
		    	i++;
		    	
		    	//Add the labels.
		    	if (x == lblXPosition
		    	  && (y % 50) == 0
		    	  && y > 0)
		    	{
		    		int probInt = 10 * y / 50;
		    		
		    		//Add a label
		    		JLabel label = new JLabel("-   " + probInt + "%   -");
		    		label.setBounds(lblXPosition, (y - (LABEL_HEIGHT/2)), LABEL_WIDTH, LABEL_HEIGHT);
		    		label.setHorizontalAlignment(SwingConstants.CENTER);
		    		panelKey.add(label);
		    	}
			}
		}
		
		keyImg.setRGB(0, 0, width, height, pixels, 0, width);
		repaint();
	}

	private Color getColorForProbability(double probability)
	{
		float hue = (float)(probability/1.2);
    	return Color.getHSBColor(hue, 1, 1);
	}
	
	@Override
	public ColourWrapper getColourWrapperForDartboard()
	{
		ColourWrapper wireframe = new ColourWrapper(DartsColour.TRANSPARENT);
		wireframe.setEdgeColour(Color.BLACK);
		return wireframe;
	}
	
	@Override
	public boolean dartboardOnTop()
	{
		//We overlay a wireframe dartboard over the shaded area
		return true;
	}

}

File: Darts/src/burlton/dartzee/code/screen/ai/VisualisationPanelDensity.kt
package burlton.dartzee.code.screen.ai

import burlton.dartzee.code.`object`.ColourWrapper
import burlton.dartzee.code.ai.AbstractDartsModel
import burlton.dartzee.code.utils.DartsColour
import burlton.dartzee.code.utils.getDistance
import burlton.desktopcore.code.bean.paint
import java.awt.Color
import java.awt.Point
import java.awt.image.BufferedImage
import javax.swing.ImageIcon
import javax.swing.JLabel
import javax.swing.SwingConstants

class VisualisationPanelDensity: AbstractVisualisationPanel()
{
    private var keyImg: BufferedImage? = null
    private val panelKey = JLabel()

    init
    {
        add(dartboard)
        add(overlay)

        val colourWrapper = ColourWrapper(DartsColour.TRANSPARENT).also { it.edgeColour = Color.BLACK }
        dartboard.paintDartboard(colourWrapper, false)

        panelKey.setBounds(0, 0, 100, 500)
        panel.add(panelKey)
        keyImg = BufferedImage(panelKey.width, panelKey.height, BufferedImage.TYPE_INT_ARGB)
        panelKey.icon = ImageIcon(keyImg)
    }

    override fun showVisualisation(hmPointToCount: Map<Point, Int>, model: AbstractDartsModel)
    {
        val centerPt = model.getScoringPoint(dartboard)

        overlayImg.paint {
            val radius = getDistance(it, centerPt)
            val probability = model.getProbabilityWithinRadius(radius)
            getColorForProbability(probability)
        }

        repaint()
    }

    override fun paintKey()
    {
        keyImg?.paint {
            val probability: Double = it.y.toDouble() / height.toDouble()
            getColorForProbability(probability)
        }

        //Add labels at 10% increments
        val lblXPosition = panel.width / 2 - LABEL_WIDTH / 2
        val yPositions = (1 until panel.height).filter { it % (height/10) == 0 }
        yPositions.forEach {
            val probInt = 10 * it / 50
            val label = JLabel("-   $probInt%   -")
            label.setBounds(lblXPosition, it - LABEL_HEIGHT / 2, LABEL_WIDTH, LABEL_HEIGHT)
            label.horizontalAlignment = SwingConstants.CENTER
            panelKey.add(label)
        }

        repaint()
    }

    private fun getColorForProbability(probability: Double): Color
    {
        val hue = (probability / 1.2).toFloat()
        return Color.getHSBColor(hue, 1f, 1f)
    }

    companion object
    {
        private const val LABEL_WIDTH = 60
        private const val LABEL_HEIGHT = 30
    }
}